langchain/document_loaders/pdf.py

```diff
@@ -381,3 +381,91 @@ def load(self) -> List[Document]:
             contents = self.clean_pdf(contents)
         metadata = {"source": self.source, "file_path": self.source}
         return [Document(page_content=contents, metadata=metadata)]
+
+
+class PDFPlumberLoader(BasePDFLoader):
+    """Loader that uses PDFPlumber to load PDF files."""
+
+    def __init__(self, file_path: str):
+        """Initialize with file path."""
+        try:
+            import pdfplumber  # noqa:F401
+        except ImportError:
+            raise ValueError(
+                "PDFPlumber package not found, please install it with "
+                "`pip install pdfplumber`"
+            )
+
+        super().__init__(file_path)
+
+    def load(self, **kwargs: Optional[Any]) -> List[Document]:
+        """Load file."""
+        import pdfplumber
+
+        doc = pdfplumber.open(self.file_path)
+        file_path = self.file_path if self.web_path is None else self.web_path
+
+        return [
+            Document(
+                page_content=page.extract_text(**kwargs).encode("utf-8"),
+                metadata=dict(
+                    {
+                        "source": file_path,
+                        "file_path": file_path,
+                        "page_number": page.page_number,
+                        "total_pages": len(doc.pages),
+                    },
+                    **{
+                        k: doc.metadata[k]
+                        for k in doc.metadata
+                        if type(doc.metadata[k]) in [str, int]
+                    },
+                ),
+            )
+            for page in doc.pages
+        ]
+
+    def annotate_and_load(
```

Comment:
@lesterpjy What do you think about decoupling the annotation from this loader into a standalone functionality? 

As a standalone functionality, it'll allow folks to use it with other ways of loading PDFs (e.g., from s3) as well as with other pdf parsers.

It'll also allow providing a better interface for some of the configuration (e.g., controlling the resolution, the file naming convention etc.).

Reply:
I think decoupling the parsing from loading is good for the loaders in https://github.com/hwchase17/langchain/pull/4356/files, but for `annotate_and_load` it can be problematic. After using pdfplumber for a bit it looks like the parsed results can be different from that of PyMuPDF, and I worry that whichever parser used by user can be unrepresented in the visualized output of pdfplumber. Hence, the coupled `annotate_and_load` to force users to only use pdfplumber as parser if they would like to utilize the visual debugger.

As for a better interface, as mentioned above with @dev2049, I am trying to figure out a way to pass **kwargs to multiple destinations. Please let me know there is a best convention. I will separate the kwargs for `extract_words`, `extract_text`, and `to_image`.

Reply:
I notice the PDF parser implementation has been merged, should I close this PR and re-implement `PDFPlumberLoader` based on the parser structure? And for the annotation functionality we could keep it as a standalone method of  `PDFPlumberLoader`? So something like this:

```python
class PDFPlumberLoader:

   def __init__(self) -> None:
      ...

   def load() -> None:
      ....
      ...parser = PDFPlumberParser(text_kwargs=kwargs)

   def save_annotated() -> None:
      ....
      ...annotated.save(save_path)
```